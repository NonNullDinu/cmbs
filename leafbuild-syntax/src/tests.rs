fn parses_without_errors(s: &str) {
    assert_eq!(crate::parser::parse(s).errors, vec![]);
}

// Those are only *parsing* tests, so as long as they make sense
// *syntactically*, they should parse.
// Even if semantically they are void of meaning.
#[test]
fn unit_to_unit_assignment() {
    parses_without_errors("() = ()\n")
}

#[test]
fn line_trailing_space() {
    parses_without_errors("() = () \n")
}

#[test]
fn if_condition() {
    parses_without_errors("if (()) {} else {}")
}

#[test]
fn if_else_if_condition() {
    parses_without_errors("if () {} else if () {} else {}")
}

#[test]
fn if_else_if_else_condition_stretched() {
    parses_without_errors(
        r#"
    if ()

    {



    }

    else if () {


    }

    else {


    }
    "#,
    )
}

#[test]
fn if_condition_in_expr() {
    parses_without_errors("() = if () {} else {} \n")
}

#[test]
fn if_condition_in_expr_stretched() {
    parses_without_errors(
        r#"
    () = if ()
    {

    }

    else
    {


    }
    "#,
    )
}

#[test]
fn declaration() {
    parses_without_errors("let x = ()\n")
}

#[test]
fn foreach_basic() {
    parses_without_errors("foreach () in () {}")
}
