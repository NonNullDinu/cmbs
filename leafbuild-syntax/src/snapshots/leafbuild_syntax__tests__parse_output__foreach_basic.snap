---
source: leafbuild-syntax/src/tests.rs
expression: "foreach () in () {}"

---
- ROOT 0..19
  - Foreach 0..19
    - "foreach" FOREACH_KW 0..7
    - " " WHITESPACE 7..8
    - Expr 8..10
      - PrimaryExpr 8..10
        - TupleExpr 8..10
          - "(" L_PAREN 8..9
          - ")" R_PAREN 9..10
    - " " WHITESPACE 10..11
    - "in" IN_KW 11..13
    - " " WHITESPACE 13..14
    - Expr 14..16
      - PrimaryExpr 14..16
        - TupleExpr 14..16
          - "(" L_PAREN 14..15
          - ")" R_PAREN 15..16
    - ExprBlock 16..19
      - " " WHITESPACE 16..17
      - "{" L_BRACE 17..18
      - "}" R_BRACE 18..19

